### PR TITLE
Nステップ先まで再帰的に想像する損失を実装した ImaginingForwardDynamicsTrainerを実装

### PR DESCRIPTION
**Pull request by Geson-anko**
## 概要

Nステップ先まで再帰的に想像し、より長期予測を可能にするための損失を実装した ImaginingForwardDynamicsTrainerを実装しました。

主に `train`メソッドの中身を確認していただけるとありがたいです。
もう少し説明が必要であればコメントお願いします。

### 参考

https://github.com/MLShukai/ami-theory/blob/main/dist/long_span_prediction_reward.pdf

## Submit前の確認項目

- [x] タイトルは一目でわかるようにし、説明文は PR を簡潔に説明するようにしましたか?
- [x] あなたの PR が、異なる変更を束ねたものではなく、ひとつのことだけを行うものであることを確認しましたか?
- [x] この PR で導入されたすべての変更点をリストアップしましたか?
- [x] PR を`make run`コマンドでローカルでテストしましたか?

### CODE REVIEW

**Comment by Geson-anko** [comment]
@myxyy 最初の想像ステップ以外は、時間軸をバッチ軸に統合することで問題を解決しました
https://github.com/MLShukai/ami/pull/287/commits/5bacf32d75c5f709e4e799c6b734a158fd9212c9

File: ami/trainers/forward_dynamics_trainer.py [hunk 1]
```diff
@@ -409,3 +410,150 @@ def load_state(self, path: Path) -> None:
...
+                        action_imaginations = action_imaginations.flatten(0, 1)  # (B', *)
+                        obs_targets = obs_targets.flatten(0, 1)  # (B', *)
+
+                    obses_next_hat_dist, next_hiddens = self.forward_dynamics(
+                        obs_imaginations, hiddens, action_imaginations
+                    )
+                    loss = -obses_next_hat_dist.log_prob(obs_targets).mean()
+                    loss_imaginations.append(loss)
+                    obs_imaginations = obses_next_hat_dist.rsample().flatten(0, 1)  # (B, T-H, *) -> (B', *)
+                    hiddens = next_hiddens.movedim(2, 1).flatten(
```
Comment (myxyy) [inline ami/trainers/forward_dynamics_trainer.py:527]: バッチサイズがどんどん増大して他と合わなくなるのでテストのimagination_length=2より大きいと動かないと思います

File: ami/trainers/forward_dynamics_trainer.py [hunk 2]
```diff
@@ -409,3 +410,150 @@ def load_state(self, path: Path) -> None:
...
+                        0, 1
+                    )  # h'_i, (B, D, T-H, *) -> (B, T-H, D, *) -> (B', D, *)
```
Comment (myxyy) [inline ami/trainers/forward_dynamics_trainer.py:529]:
```suggestion
                    if i > 0:
                        hiddens = next_hiddens[:,:,0,:]
                    else:
                        hiddens = next_hiddens.movedim(2, 1).flatten(
                            0, 1
                        )  # h'_i, (B, D, T-H, *) -> (B, T-H, D, *) -> (B', D, *)
```
になるかなと思います

Comment (Geson-anko) [reply ami/trainers/forward_dynamics_trainer.py]: @myxyy `(*batch, dim), (*batch, depth, dim) -> (*batch, dim), (*batch, depth, dim)`の方を用いて修正しました。

https://github.com/MLShukai/ami/pull/287/commits/a8796a15797a5e561d3584196e220eceed0bb123

**Review by myxyy — APPROVED** [review]
